experimental aggregator module

grondo opened the pull request:
For comment only. The purpose of this experiment is to play with the reduction/aggregation interface for possible use with wrexec task exit status. The PR is posted early to assist with design of generic reduction handle, get early feedback,e tc.

The current `aggregator` module reduces set of `id, value` pairs into a kvs directory with the contents:
```
key.max
key.min
key.count
key.values.
```
Where the `key.values` directory has one entry per value, with the contents of each values entry a nodelist of ids that reported that value.

A `flux-aggregate` test command is also available for testing aggregation of a value across a session (currently version assumes one value per rank, i.e. run `flux aggregate <key> <value>` once per rank) The `flux-aggregate` command running on rank 0 waits for the result in the given key and reports the result e.g:
```
$ flux exec -r 0-127 sh -c 'flux aggregate test1 $(flux getattr size)'
[0-127]: 128

$ flux exec -r 0-127 sh -c 'flux aggregate testkey $(flux getattr tbon.level)' | sort -k2,2n
0: 0
[1-2]: 1
[3-6]: 2
[7-14]: 3
[15-30]: 4
[31-62]: 5
[63-126]: 6
127: 7

$ flux kvs dir -r testkey
testkey.min = 0
testkey.max = 7
testkey.count = 128
testkey.values.7 = 127
testkey.values.1 = [1-2]
testkey.values.5 = [31-62]
testkey.values.6 = [63-126]
testkey.values.3 = [7-14]
testkey.values.4 = [15-30]
testkey.values.2 = [3-6]
testkey.values.0 = 0
```

The module is designed to have a single service `aggregator.push`, which accepts a JSON representation of an "aggregate" and either creates a new aggregate and hashes it by kvs key, or merges the incoming JSON with an existing matching aggregate. 

When a new aggregate is seen, this starts a timer, after which the current aggregate is forwarded upstream and its state is immediately destroyed.

On rank 0, a timer is not used, instead the aggregate is written to the kvs once `aggregate->count == aggregate->total`.

grondo commented:
The current implementation does not use a reduction handle. A reduction handle could be used, but would require extra code to either

 * Keep all aggregate state in JSON objects, with functions to merge two JSON objects for the same reduction

or

 * Convert incoming JSON directly to a `struct aggregate *` then push aggregates directly onto reduction handle queue for later merging.

I'd actually like to experiment with the first option above in a separate branch.

garlick commented:
> Keep all aggregate state in JSON objects, with functions to merge two JSON objects for the same reduction

So you'd create a `flux_reduce_t` object per key?

This approach is better aligned with the design (flawed as it may be).  "items" were intended to be the same type regardless of whether they were single or composite (hence the "itemweight" method).

grondo commented:
> So you'd create a flux_reduce_t object per key?

Yeah, and (if the current design is followed anyway), the flux_reduce_t objects would be created "on demand" (if `key` is not already cached) and be destroyed (or recycled) after each `sink` or `forward`.

> This approach is better aligned with the design (flawed as it may be). "items" were intended to be the same type regardless of whether they were single or composite (hence the "itemweight" method).

Yeah, this is the eventual conclusion I came to in a roundabout way above. This by design prevents the reduction of incoming messages into an existing data structure or state. You either have to convert messages to a data structure "item" to reduce them, or keep items as messages (or json) and convert back and forth to/from data structures on the fly to reduce (perhaps not for the simplest of reductions). Either way it is a bit of extra code. (but perhaps not a bad thing in order to abstract the reduction semantics itself -- though in this specific case that is just a timer, so it is a hard sell that the reduction handle offered any simplification.)

grondo commented:
Still playing with ideas and interfaces here...
Just pushed support for adding a rank-local `forward-count` to an aggregator message, to be used when the number of items expected at any rank is known (e.g. in this test case we are sending exactly one item per rank, so we can base fwd-count off the `tbon.descendants` attribute)

e.g.
```
$ time flux exec -r 0-1919 bash -c 'flux aggregate -t 100 -c $((1+`flux getattr tbon.descendants`)) testkey $(flux getattr tbon.level)'
[1-2]: 1
[7-14]: 3
[31-62]: 5
[127-254]: 7
[511-1022]: 9
[1023-1919]: 10
[255-510]: 8
[63-126]: 6
[15-30]: 4
[3-6]: 2
0: 0

real	0m0.974s
user	0m0.432s
sys	0m0.065s

```

At this point, not much faster than the default (and there's a chance I got something wrong )

```
$ time flux exec -r 0-1919 bash -c 'flux aggregate testkey $(flux getattr tbon.level)'
[1023-1919]: 10
[511-1022]: 9
[255-510]: 8
[127-254]: 7
[3-6]: 2
[7-14]: 3
[15-30]: 4
[63-126]: 6
[31-62]: 5
[1-2]: 1
0: 0

real	0m1.220s
user	0m0.427s
sys	0m0.064s

```


One problem with the current approach is that the fwd-count must be per-rank, so if you generate aggregates on a subset of ranks, we have to have some way to notify downstream ranks of their own fwd-count.

However, at this point this feature allows us to compare timer vs count-based reduction.

grondo commented:
Oh, I just realized the timings of these tests above may be somewhat influenced by `flux exec` slowness . I should probably add a barrier to the `flux aggregate` and do the timing on rank 0 from barrier exit to completion. Sorry about that....

dongahn commented:
@grondo: 

Thank you for posting this early!  I believe that reduction is a key to high scalability and using a KVS key to isolate each distinct reduction makes a lot of sense to me. 

One high level discussion which I think would help me a lot is actually to make some progress on this comment of yours:

> The PR is posted early to assist with design of generic reduction handle, get early feedback, etc.

When I looked at `flux_reduce_t`, it offers good primitive operations but doesn't have an abstraction to manage isolation and I thought a KVS key can offer an isolation based on a name, and also a good place to land the reduced data. I can image we can have MPI-style All_Reduce to distribute the reduced data back to the senders but I thought KVS-based reduction has a large set of its own use cases. In short, I'm all for creating a generic reduction interface around KVS. 

In terms of generic reduction interface design, thought, I have a few questions to consider:

1. Do we want to provide a set of common reduction operators like what MPI provides (e.g., max, min, sum... over both integer and floating points)?

2. Do we want to provide extended common reduction operators beyond what MPI provides (e.g., list compression -- like {1,2,3,4,6} to 1-4,6)? Are there anything else beyond list compression operators?

3. Do we want to provide these as a module interface (e.g., `aggregator_reduce ()`) as well so that one can use this interface instead of a raw rpc?

4. What should be the semantics of our abstraction in particular in terms of synchronous vs. asynchronous or both. 

5.  Do we want to go beyond this and allow users to provide a user-defined reduction function?

I think Item 1 - 4 is a matter of collecting collective wisdom and use cases to decide what the right design choices should be.

Item 5 goes a bit deeper because one has to determine a mechanism by which an aggregator module to load in the user-defined function. If the mechanism should be via a Lua script or shared library, it would be better to keep interface data parameters as data-agnostic like JSON or `void *`

grondo commented:
> Do we want to provide a set of common reduction operators like what MPI provides (e.g., max, min, sum... over both integer and floating points)?

I was thinking the "numeric" reduction could eventually just use double as the base type, and allow either floating point or integer values. However, this breaks my current design for using values as keys (can't store a floating point number as a key in our kvs unfortunately), so I didn't go there yet.

> Do we want to provide extended common reduction operators beyond what MPI provides (e.g., list compression -- like {1,2,3,4,6} to 1-4,6)? Are there other common operators beyond list compression operators?

I think I'm a little lost on what you mean by a "reduction operator" (not coming from a background of other reduction schemes). This current experiment is simplest solution to address a current use case, the reduction of a set of likely overlapping values from a list of ids into a single kvs_commit. Since we're actually reducing a pair of values do any of your operators apply?

> Do we want to provide these as a module interface (e.g., aggregator_reduce ()) as well so that one can use this interface instead of a raw rpc?

It doesn't feel like it is time to go there yet, but possibly in the future? I worry about creating a static API interface and thus restricting the implentation, when a "message API" is much more flexible, easier to implement backwards compatibility, etc. However, I suppose for common use cases a simplified API might be useful eventually.

> What should be the semantics of our abstraction in particular in terms of synchronous vs. asynchronous? Do we want to provide both semantics?

I'm not sure what you mean here? For synchronous, do you mean something like the `aggregator.push` RPC continuation doesn't complete until after reduction is finished? If so, synchronous behavior can be implemented by installing a watch on destination kvs key `key.count` value. I'm not exactly a fan of creating synchronous-only interfaces in event driven architecture though...

> 5.  Do we want to go beyond this and allow users to provide a user-defined reduction function?

This will be experiment 2, which I'll be working on now. Experiment 2 will be composed of a reduction module which will allow Lua scripts to be registered that perform the "reduce" function on JSON objects.
(perhaps also the "sink" operation can be part of the Lua script)
We'll see how that one goes...

grondo commented:
For fun, I added a barrier to the start of the `flux-aggregate`, and delayed start of the reduction until after the barrier is complete. The test now prints timing for the barrier and the subsequent reduction to kvs.

I did some tests on 92 nodes of hype, 32 brokers per node:

```sh
$ flux getattr size
2944

# Default timer based reduction:
$ flux exec -r 0-2943 bash -c 'flux aggregate testkey $(flux getattr size)'
aggregate: barrier took 1.006s
[0-2943]: 2944
aggregate: Took 0.321s

# Explicit count based reduction:
$ flux exec -r 0-2943 bash -c 'flux aggregate -t 100  -c $((1+$(flux getattr tbon.descendants))) testkey $(flux getattr size)'
aggregate: barrier took 1.394s
[0-2943]: 2944
aggregate: Took 0.256s

# No reduction (timeout set to 0.0)
$ flux exec -r 0-2943 bash -c 'flux aggregate -t 0.0 testkey $(flux getattr size)'
aggregate: barrier took 0.989s
[0-2943]: 2944
aggregate: Took 2.026s

```

garlick commented:
Nice!

SteVwonder commented:
@grondo: what are the semantics of `-t`?  I presume it is the time between when the first aggregate is received and when the current aggregate is pushed upstream.

dongahn commented:
> I think I'm a little lost on what you mean by a "reduction operator" (not coming from a background of other reduction schemes). This current experiment is simplest solution to address a current use case, the reduction of a set of likely overlapping values from a list of ids into a single kvs_commit. Since we're actually reducing a pair of values do any of your operators apply?

Yes, as far as one can express a reduction operator as `value = value op value,` this should work, if this makes sense.

I am mostly coming from the MPI-based reduction lingo. Like: https://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-1.1/node78.htm

If you look at the current implementation from this prospective, the current aggregator has three operators over integers (one integer per participants): max, min and compressed concatenation.

dongahn commented:
> It doesn't feel like it is time to go there yet, but possibly in the future? I worry about creating a static API interface and thus restricting the implentation, when a "message API" is much more flexible, easier to implement backwards compatibility, etc. However, I suppose for common use cases a simplified API might be useful eventually.

Yes, this sounds right. We can have a flexible mechanic in place first and add an user-friendly interface on top of that.

dongahn commented:
> I'm not sure what you mean here? For synchronous, do you mean something like the aggregator.push RPC continuation doesn't complete until after reduction is finished? If so, synchronous behavior can be implemented by installing a watch on destination kvs key key.count value. I'm not exactly a fan of creating synchronous-only interfaces in event driven architecture though...

I think this will matter more when we have a higher-level module-interface abstraction. It feels we will probably need both async and sync versions of such interface.

dongahn commented:
> This will be experiment 2, which I'll be working on now. Experiment 2 will be composed of a reduction module which will allow Lua scripts to be registered that perform the "reduce" function on JSON objects.
(perhaps also the "sink" operation can be part of the Lua script)
We'll see how that one goes...

This sounds good. I'm not sure allowing sink as part of a user-defined reduction is a good idea though, strictly from the perspective of separation of concerns. Users probably want to describe how values should be reduced and flux's aggregation infrastructure calls them back at each reduction step, taking care of data motions under the hood... At least most of distributed/parallel reductions use this pattern.

grondo commented:
>  So -t 0.0 devolves into the original case where rank 0 will receive N messages (i.e. no aggregation is done).

Correct. Though in the module I special cased `0.` so that the aggregate is forwarded immediately upon being received, without going through the timer. (Otherwise I *think* multiple messages per reactor loop would actually be aggregated, and I wanted to test worst case)

dongahn commented:
> Yes, as far as one can express a reduction operator as value = value op value, this should work, if this makes sense.

> I am mostly coming from the MPI-based reduction lingo. Like: https://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-1.1/node78.htm

> If you look at the current implementation from this prospective, the current aggregator has three operators over integers (one integer per participants): max, min and compressed concatenation.


@grondo: Sorry for my sporadic responses. I'm in transition...

Idea: What if we allow users to name the reduction operator or operators to use as well? IOW, a vector of \<key, value, named operator(s)\> instructs the aggregator how the values should be reduced. For example, 

- \<mykey, value, max\> can results in 
`mykey.max.value = max of all the input values`
- \<mykey, value, sum\> 
`mykey.sum.value = sum of all the input values`
- \<mykey, value, compressed-concat\>
`mykey.compressed-concat.value = compressed list of integers` or your current representation

And one can also compose these operators like
- \<mykey, value, max,min,compressed-concat\> which can result in
```
mykey.max.value = max
mykey.min.value = min
mykey.compressed-concat.value = compressed list
```

I guess some of these well know operators can be built-in with respect to some value types so the users can simply name them. If they want to register they their own, they can also "name" it, and register a Lua script for it before using it? Then, the underlying aggregator pluming will invoke the list of reduction functions one by one for each reduction step?

grondo commented:
> Idea: What if we allow users to name the reduction operator or operators to use as well? IOW, a vector of <key, value, named operator(s)> instructs the aggregator how the values should be reduced.

Thanks @dongahn, that idea does seem to have merit !

I guess I was thinking that, for Flux, reduction is over complex types (messages) not simple values like in the MPI case. I'm wondering if all possible complex type reductions could even be composed of the basic numeric operators like count, max, min, or even compressed-concatenation, though I can't think of a good example right now (actually perhaps MPIR_Procdesc is a good example?)

A perhaps simpler idea would be to have a named type system with an interface to register reduction functions for these named types. The type would be a field in the reduction message, or perhaps even appended to the topic string, e.g. `reduce.push.integer`.

This elevates `reduce()` function to work on abstract types, instead of only integer or numeric values, and I think thus could be more flexible and usable for real-world use cases like collection of task exit codes, procdesc values, stdio etc.

dongahn commented:
@grondo: completely agreed that we will need to be able to build a reduction over complex types. In fact, MPI's user-defined reduction function requires `MPI_Datatype` to be passed, so that the reduction operator can appropriately interpret `void *` input vector as a vector of the user-defined (or primitive) data type. 

I like your proposal to embed the data type info into the input itself and associate a reduction operator for it. Just extending that, though, we may have use cases where we need more than one type of reduction operator for that particular type. Integer would be a simple case where there are many reduction operators that may be needed and unless you want a single reduction function does all of the reduction operations at once, you may find it hard to extend it with another operator we cannot even conceive now?

Do you find a merit to embed a reduction operator name in the input itself as well? If there is only one reduction operator registered, one can skip specifying that and that system becomes the same as what your proposed?

grondo commented:
> Just extending that, though, we may have use cases where we need more than one type of reduction operator for that particular type. 

Oh, I was suggesting the `type` field be the type of reduction (which applies to only a given message type). So you could have multiple reduction functions that all operate on integers (e.g. `integer-sum`, `integer-stats`...) (Though if `stats` included `min, max, sum, count, etc` it seems like you might always just want to use the composite and not need individual reductions for all the others)

I think that is what you mean. Apologies if I'm missing your use case.

dongahn commented:
@grondo: okay thanks! I think this proposal should cover most of the basis! If and when you make some progress for this I'd happy to play with it and be a 2nd pair of eyes. I am particularly interested in what interface signiture you will require from a user defined reduce operator.

grondo commented:
As an experiment I added support to wrexecd to optionally push task exit messages through this current version of the aggregator module, and unsurprisingly there is a big benefit. I'm not getting consistent launch times on opal/jade (NFS slowness?), however, the most obvious improvement is in the time it takes the `flux-wreckrun` script to gather and return the maximal exit code. (one kvs_get instead of 1860 synchronous gets is an obviously huge improvement)

These runs were just across 93 nodes
```
$ flux wreckrun -v -n1860  /bin/true
wreckrun: 0.005s: Sending LWJ request for 1860 tasks (cmdline "/bin/true")
wreckrun: 0.060s: Registered jobid 15
wreckrun: 0.061s: State = reserved
wreckrun: Allocating 1860 tasks across 93 available nodes..
wreckrun: tasks per node: node[0-92]: 20
wreckrun: 0.161s: Sending run event
wreckrun: 8.829s: State = starting
wreckrun: 9.394s: State = running
wreckrun: 9.394s: State = complete
wreckrun: tasks [0-1859]: exited with exit code 0
wreckrun: 14.637s: All tasks completed successfully.
$ flux wreckrun -o aggregate-task-exit -v -n1860  /bin/true
wreckrun: 0.002s: Sending LWJ request for 1860 tasks (cmdline "/bin/true")
wreckrun: 0.055s: Registered jobid 16
wreckrun: 0.056s: State = reserved
wreckrun: Allocating 1860 tasks across 93 available nodes..
wreckrun: tasks per node: node[0-92]: 20
wreckrun: 0.156s: Sending run event
wreckrun: 7.510s: State = starting
wreckrun: 7.825s: State = running
wreckrun: 7.825s: State = complete
wreckrun: tasks [0-1859]: exited with exit code 0
wreckrun: 7.908s: All tasks completed successfully.

```

garlick commented:
Excellent result!  We should get this in and rerun CTS-1 numbers.

garlick commented:
> We should get this in and rerun CTS-1 numbers.

Er I meant more formally, since that's exactly what you did here.
(Should have just clicked the emoticon :-)

dongahn commented:
Because this will improve kvs performance as job sizes increase, there is a good chance this will improve the current scheduler throughput issue. I'd like to have this available in the master soon as well and rebase my scheduling perf/scalability result. Thank you @grondo!

grondo commented:
Ok, I've pushed what I have currently for lwj task exit code aggregation. The default in this branch is to now use the aggregator module by default unless `-o no-aggregate-task-exit` is passed as a job option. The nodeid 0 of the job will wait until all task exit codes make it to the kvs before marking the job completed.

For this to become a merge candidate there is a little more work required -- tests, error handling for the aggregator module, etc.

grondo commented:
Sorry about the noise. Recent updates in this PR change the result of the aggregation from a directory hierarchy in KVS to a single JSON object of the same format as the aggregator JSON messages. Additionally an `aggregator.abort` event is added that's used to abort an aggregate that fails to put or commit to the KVS.

If this is getting close to acceptable I would squash out intermediate history.

I still may also want to have a timeout on the rank 0 node to let result trickle in to KVS when aggregation  is slow, instead of always waiting for aggregate count to reach the expected total.

dongahn commented:
> Sorry about the noise. Recent updates in this PR change the result of the aggregation from a directory hierarchy in KVS to a single JSON object of the same format as the aggregator JSON messages. 

Sounds good to me and should be more performant. Was the design decision due to performance or was there something else to it?

> Additionally an aggregator.abort event is added that's used to abort an aggregate that fails to put or commit to the KVS.

Just to clarify, would the main user of this event is the ones that started aggregation so that they can asynchronously be notified of an abort/success?

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+
+static struct aggregate_entry * aggregate_entry_create (void)
+{
+    struct aggregate_entry *ae = xzmalloc (sizeof (*ae));
```
consider using calloc here and checking return value

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+
+    json_object_object_foreachC (entries, i) {
+        int64_t val = json_object_get_int64 (i.val);
+        aggregate_push (ag, val, i.key);
```
check for NULL return

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+    }
+    if ((msg = flux_event_encode (topic, "{ }")) == NULL) {
+        flux_log_error (h, "flux_event_encode");
+        return;
```
leaks 'topic'

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+        flux_log_error (h, "flux_event_encode");
+
+    flux_msg_destroy (msg);
+}
```
leaks topic

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+            t = .250;
+        flux_log (h, LOG_INFO, "sink: %s: retry  in %.3fs", ag->key, t);
+        /* On failure, retry just once, then abort */
+        w = flux_timer_watcher_create (flux_get_reactor (h),
```
check for null return

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+{
+    flux_t h = ctx->h;
+
+    struct aggregate *ag = xzmalloc (sizeof (*ag));
```
consider switching to cmalloc(), check return for NULL, drop the memset

grondo replied:
it occurs to me if `xzmalloc` is contraindicated for modules maybe we should wrap it in some sort of warning. Not sure if this is even possible though since we don't require modules to load `module.h` or anything... Just an idea to throw out there

garlick replied:
Yeah, or we should decide it's allowed, or figure out how to wire in a more appropriate `oom()` handler.  In this case there was already an ENOMEM path for `zlist_new()` so it stuck out a bit.

dongahn replied:
sorry guys, I didn't know about `xzmalloc()` being contraindicated. We have lots of code using it in sched as well.

garlick replied:
Exiting from a module is contraindicated, because it will cause the broker to exit.  However, managing ENOMEM as a transient error may be a losing proposition, and fixing all the other places (e.g. shortjson) where this occurs is a big job.

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+    struct aggregate *ag = xzmalloc (sizeof (*ag));
+    memset (ag, 0, sizeof (*ag));
+    ag->ctx = ctx;
+    ag->key = strdup (key);
```
check for NULL return

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+static void aggregator_destroy (struct aggregator *ctx)
+{
+    if (ctx) {
+        if (ctx->aggregates)
```
no need to test for NULL before callling

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+    l = strtoul (s, &p, 10);
+    if (*p != '\0' || errno != 0) {
+        flux_log_error (h, "flux_attr_get (%s) = %s", attr, s);
+        return (-1);
```
failure of -1 is a valid return?  also callers aren't checking.

grondo replied:
Working through these now, but this one I couldn't follow, sorry.

Are you asking about `attr_get_int` return `-1` on failure? The two callers do seem to check this, but I have to assume I missed your comment here: (-1 indicates error from this function)
```C
static double timer_scale (flux_t h)
{
    long level, maxlevel;
    if (((level = attr_get_int (h, "tbon.level")) < 0) ||
        ((maxlevel = attr_get_int (h, "tbon.maxlevel")) < 0)) {
        return (1.);
    }
    return (maxlevel - level + 1.);
}
```

garlick replied:
Better ignore that one, I don't know _what_ I thought I saw now. :-)

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+    struct aggregator * ctx = xzmalloc (sizeof (*ctx));
+    ctx->h = h;
+    if (flux_get_rank (h, &ctx->rank) < 0) {
+        flux_log_error (h, "zhash_new");
```
wrong error message

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+        flux_log_error (h, "zhash_new");
+        goto error;
+    }
+    ctx->aggregates = zhash_new ();
```
check for null return

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+    if (fwd_count > 0)
+        ag->fwd_count = fwd_count;
+
+    rc = aggregate_push_json (ag, in);
```
goto done if rc < 0?

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+        if (ag->count == ag->total
+            || ag->count == ag->fwd_count
+            || timeout == 0.)
+            aggregate_flush (ag);
```
check for failure

garlick commented on src/modules/aggregator/aggregator.c:
```diff
@@ -0,0 +1,534 @@
+            aggregate_flush (ag);
+    }
+    else if (ag->count == ag->total)
+        aggregate_try_sink (h, ag);
```
check for failure

garlick commented on src/modules/wreck/wrexecd.c:
```diff
@@ -666,9 +650,70 @@ void prog_ctx_destroy (struct prog_ctx *ctx)
+        goto out;
+    }
+    if (!(intp = zhash_lookup (ctx->completion_refs, ref))) {
+        intp = xzmalloc (sizeof (*intp));
```
switch to calloc() and check return?

garlick commented on src/modules/wreck/wrexecd.c:
```diff
@@ -677,6 +722,7 @@ struct prog_ctx * prog_ctx_create (void)
 
     memset (ctx, 0, sizeof (*ctx));
     ctx->options = zhash_new ();
+    ctx->completion_refs = zhash_new ();
```
check for NULL return

garlick commented on src/modules/wreck/wrexecd.c:
```diff
@@ -666,9 +650,70 @@ void prog_ctx_destroy (struct prog_ctx *ctx)
 
     free (ctx->cores_per_node);
 
+    if (ctx->options)
```
no need to check for null before calling zhash_destroy

garlick commented:
In some of the comments above I tagged a line near the problem and the context github decided to include was a bit lacking.  Should be obvious if you look at the source - sorry about that!

grondo commented:
Sorry I should have mentioned I hadn't made a pass through this to check those kinds of errors and was looking for more generic feedback. Thanks for the line notes though, @garlick, should make it much easier!

grondo commented:
> Sounds good to me and should be more performant. Was the design decision due to performance or was there something else to it?

I assume you are talking about the decision to use a single key instead of a directory of keys, e.g. `result.min` `result.max`. This choice was suggested by garlic to reduce the total number of keys, especially when running single task jobs (in which case there would be slightly more total keys in the aggregation case vs previous method)

> Just to clarify, would the main user of this event is the ones that started aggregation so that they can asynchronously be notified of an abort/success?

Yes, as an example `wrexecd` is a user of this service and the rank 0 wrexecd does not mark the job as `"complete"` until either the aggregate `lwj.*.exit_status` has made it to the kvs (via `kvs_watch`) or the aggregate is aborted. This will allow jobs that fail to write to kvs during aggregation to still complete, though exit code is lost.

grondo commented:
Ok, I've rebased and squashed this PR to make it a bit easier to digest. I think I've addressed most of @garlick's comments above. I still have to take one more pass to fix further dumb mistakes, ENOMEM handling, and leaks.

Some other things that might need discussion before this goes in:
 - `flux aggregate` is added here, but I'm not sure it is too useful outside of testing (?) It could be used only to aggregate a set of values, one per rank, across the *whole* session at this point.
 - Lua bindings needed a way to decode exit status from `wait`, the `flux.exitstatus` function added in this PR isn't satisfying.
 - Currently aggregator will only `sink()` to kvs when all aggregated values have been received on rank 0. It might be nice, especially for wreckrun to be able to record tasks that have "exited early"

However, this aggregator module should only be considered a temporary stand-in until we get more generic lua-based reduction.

garlick commented:
The items you listed seem like things that could be tweaked later if need be?

I'm not hard over on keeping experimental stuff out of the "command space" but there are a few test executables built under `t/` if that's a better way to go, e.g. see `t/kvs`

grondo commented:
> I'm not hard over on keeping experimental stuff out of the "command space" but there are a few test executables built under t/ if that's a better way to go, e.g. see t/kvs

Yeah, and that is where we keep the test helper scripts too (like the new `kvs-get-ex.lua`). However, it is nice enough to be able to `flux exec -r 0-N flux aggregate ...` that perhaps it is fine to leave the command for now -- we probably won't have another, similarly named command, so there might be no problem.

garlick commented:
Works for me!

garlick commented:
Ready to merge this one?

grondo commented:
Running valgrind and going to push some leak fixes a bit later. Then probably ready...

grondo commented:
Ok, pushed some memleak fixes and now this could be a candidate to go in.

garlick commented:
OK this is great.  Merging.
